2023.1.0 regression: `fs.ls(..., versions=True)` raises when subdirectories exist

`fs.ls(..., versions=True)` crashes whenever its output contains a directory. This is probably related to #509.

To repro:
```python
def test_ls_versions(gcs):
    assert 'nested' in gcs.ls(TEST_BUCKET, versions=True)
```
Traceback:
```
gcs = <gcsfs.core.GCSFileSystem object at 0x102d0d100>

    def test_ls_versions(gcs):
>       assert 'nested' in gcs.ls(TEST_BUCKET, versions=True)

gcsfs/tests/test_core.py:1211: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../.pyenv/versions/3.8.13/envs/gcsfs-dev/lib/python3.8/site-packages/fsspec/asyn.py:114: in wrapper
    return sync(self.loop, func, *args, **kwargs)
../../../.pyenv/versions/3.8.13/envs/gcsfs-dev/lib/python3.8/site-packages/fsspec/asyn.py:99: in sync
    raise return_result
../../../.pyenv/versions/3.8.13/envs/gcsfs-dev/lib/python3.8/site-packages/fsspec/asyn.py:54: in _runner
    result[0] = await coro
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <gcsfs.core.GCSFileSystem object at 0x102d0d100>, path = 'gcsfs_test', detail = False, prefix = '', versions = True, kwargs = {}, out = []
entry = {'bucket': 'gcsfs_test', 'name': 'gcsfs_test/nested', 'size': 0, 'storageClass': 'DIRECTORY', ...}

    async def _ls(self, path, detail=False, prefix="", versions=False, **kwargs):
        """List objects under the given '/{bucket}/{prefix} path."""
        path = self._strip_protocol(path).rstrip("/")
    
        if path in ["/", ""]:
            out = await self._list_buckets()
        else:
            out = []
            for entry in await self._list_objects(
                path, prefix=prefix, versions=versions
            ):
                if versions:
                    out.append(
                        {
                            **entry,
>                           "name": f"{entry['name']}#{entry['generation']}",
                        }
                    )
E                   KeyError: 'generation'

gcsfs/core.py:842: KeyError
```

--- a/gcsfs/core.py
+++ b/gcsfs/core.py
@@ -835,15 +835,10 @@ class GCSFileSystem(AsyncFileSystem):
             for entry in await self._list_objects(
                 path, prefix=prefix, versions=versions
             ):
-                if versions:
-                    out.append(
-                        {
-                            **entry,
-                            "name": f"{entry['name']}#{entry['generation']}",
-                        }
-                    )
-                else:
-                    out.append(entry)
+                if versions and "generation" in entry:
+                    entry = entry.copy()
+                    entry["name"] = f"{entry['name']}#{entry['generation']}"
+                out.append(entry)
 
         if detail:
             return out

--- a/gcsfs/tests/test_core.py
+++ b/gcsfs/tests/test_core.py
@@ -1205,6 +1205,7 @@ def test_ls_versioned(gcs_versioned):
     assert versions == {
         entry["name"] for entry in gcs_versioned.ls(dpath, detail=True, versions=True)
     }
+    assert gcs_versioned.ls(TEST_BUCKET, versions=True) == ["gcsfs_test/tmp"]
 
 
 def test_find_versioned(gcs_versioned):

